File: Java/Flyweight/HumansVsOrcs/src/com/ericgtkb/PlayerCreator.java
package com.ericgtkb;

import java.util.HashMap;
import java.util.Map;

public class PlayerCreator {
    private static Map<String, Player> playerMap = new HashMap<>();

    public static Player getPlayer(String playerType) throws Exception {
        if (playerMap.containsKey(playerType)) {
            return playerMap.get(playerType);
        } else {
            Player player = null;

            switch (playerType) {
                case "human":
                    System.out.println("Creating a human...");
                    player = new Human();
                    playerMap.put(playerType, player);
                    break;
                case "orc":
                    System.out.println("Creating an orc...");
                    player = new Orc();
                    playerMap.put(playerType, player);
                    break;
                default:
                    throw new Exception("Unknown player type...");
            }
            return player;
        }
    }
}

File: C++/Flyweight/HumansVsOrcs/player_creator.cc
#include <iostream>
#include "player_creator.h"

std::unordered_map<std::string, std::unique_ptr<Player>>  PlayerCreator::player_map{};

Player *PlayerCreator::get_player(const std::string& player_type) {
    if (player_map.find(player_type) != player_map.end()) {
        return player_map[player_type].get();
    } else {
        Player* player {nullptr};
        if (player_type == "human") {
            std::cout << "Creating a human...\n";
            player_map[player_type] = std::make_unique<Human>();
            player = player_map[player_type].get();
        } else if (player_type == "orc") {
            std::cout << "Creating an orc...\n";
            player_map[player_type] = std::make_unique<Orc>();
            player = player_map[player_type].get();
        } else {
            std::cout << "Unknown player type...\n";
        }
        return player;
    }
}
